plenum/server/consensus/message_request/message_req_3pc_service.py

```diff
@@ -0,0 +1,135 @@
+import logging
+from typing import Dict, List
+
+from plenum.common.constants import LEDGER_STATUS, PREPREPARE, CONSISTENCY_PROOF, \
+    PROPAGATE, PREPARE, COMMIT
+from plenum.common.event_bus import InternalBus, ExternalBus
+from plenum.common.exceptions import IncorrectMessageForHandlingException
+from plenum.common.messages.internal_messages import Missing3pcMessage, CheckpointStabilized, ViewChangeStarted
+from plenum.common.messages.node_messages import MessageReq, MessageRep, Ordered
+from plenum.common.metrics_collector import measure_time, MetricsName, NullMetricsCollector
+from plenum.common.router import Subscription
+from plenum.common.types import f
+from plenum.common.util import compare_3PC_keys
+from plenum.server.consensus.consensus_shared_data import ConsensusSharedData
+from plenum.server.consensus.message_request.message_handlers import PreprepareHandler, PrepareHandler, CommitHandler
+from stp_core.common.log import getlogger
+
+
+class MessageReq3pcService:
+    def __init__(self,
+                 data: ConsensusSharedData,
+                 bus: InternalBus,
+                 network: ExternalBus,
+                 metrics=NullMetricsCollector()):
+        self._logger = getlogger()
+        self._data = data
+        self._bus = bus
+        self._subscription = Subscription()
+        self._subscription.subscribe(bus, Missing3pcMessage, self.process_missing_message)
+        self._subscription.subscribe(bus, Ordered, self.process_ordered)
+        self._subscription.subscribe(bus, ViewChangeStarted, self.process_view_change_started)
+        self._subscription.subscribe(bus, CheckpointStabilized, self.process_checkpoint_stabilized)
+
+        self._network = network
+        self._subscription.subscribe(network, MessageReq, self.process_message_req)
+        self._subscription.subscribe(network, MessageRep, self.process_message_rep)
+
+        self.metrics = metrics
+        self.handlers = {
+            PREPREPARE: PreprepareHandler(self._data),
+            PREPARE: PrepareHandler(self._data),
+            COMMIT: CommitHandler(self._data),
+        }
+
+    @measure_time(MetricsName.PROCESS_MESSAGE_REQ_TIME)
+    def process_message_req(self, msg: MessageReq, frm):
+        # Assumes a shared memory architecture. In case of multiprocessing,
+        # RPC architecture, use deques to communicate the message and node will
+        # maintain a unique internal message id to correlate responses.
+        msg_type = msg.msg_type
```

Comment:
What if this is unknown/unsupported `msg_type` (invalid one or message request of Ledger Statuses for which we don't have registered handlers)?

Reply:
fixed